wandb/data_types.py

```diff
@@ -115,18 +158,359 @@ def from_keras(cls, model):
         layers = model.layers
         for i in range(len(layers)):
             node = Node.from_keras(layers[i], relevant_nodes)
-            graph.nodes.append(node)
+            graph.add_node(node)
+
+        return graph
 
+    @classmethod
+    def hook_torch(cls, model, criterion=None):
+        graph = cls("torch")
+        graph.hook_torch_modules(model, criterion)
         return graph
 
+    def create_forward_hook(self, name, modules):
+        graph = self
+
+        def after_forward_hook(module, input, output):
+            if id(module) in modules:
+                return
+            modules.add(id(module))
+            if not isinstance(output, tuple):
+                output = (output,)
+            parameters = [(pname, list(param.size()))
+                          for pname, param in module.named_parameters()]
+            node = Node(
+                id=id(module),
+                name=name,
+                class_name=str(module),
+                output_shape=[list(o.shape) for o in output],
+                parameters=parameters,
+                num_parameters=[reduce(mul, size)
+                                for (pname, size) in parameters]
+            )
+            graph.nodes_by_id[id(module)] = node
+            for param in module.parameters():
+                graph.nodes_by_id[id(param)] = node
+            graph.add_node(node)
+            if not graph.criterion_passed:
+                graph.criterion = output[0].grad_fn
+        return after_forward_hook
+
+    def hook_torch_modules(self, module, criterion=None, prefix=None):
+        torch = util.get_module("torch", "Could not import torch")
+        hooks = []
+        modules = set()
+        layers = 0
+        graph = self
+        if criterion:
+            graph.criterion = criterion
+            graph.criterion_passed = True
+
+        for name, sub_module in module._modules.items():
+            name = name or str(layers)
+            if prefix:
+                name = prefix + "." + name
+            layers += 1
+            if not isinstance(sub_module, torch.nn.Module):
+                # TODO: Why does this happen?
```

Comment:
Hmm... maybe there's a better superclass to use. Is it maybe the built-in modules that break this?